target_chains/ethereum/contracts/contracts/random/PythRandom.sol

```diff
@@ -0,0 +1,324 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.0;
+
+import "./PythRandomState.sol";
+import "./PythRandomErrors.sol";
+import "./PythRandomEvents.sol";
+
+// PythRandom implements a secure 2-party random number generation procedure. The protocol
+// is an extension of a simple commit/reveal protocol. The original version has the following steps:
+//
+// 1. Two parties A and B each draw a random number x_{A,B}
+// 2. A and B then share h_{A,B} = hash(x_{A,B})
+// 3. A and B reveal x_{A,B}
+// 4. Both parties verify that hash(x_{A, B}) == h_{A,B}
+// 5. The random number r = hash(x_A, x_B)
+//
+// This protocol has the property that the result is random as long as either A or B are honest.
+// Thus, neither party needs to trust the other -- as long as they are themselves honest, they can
+// ensure that the result r is random.
+//
+// PythRandom implements a version of this protocol that is optimized for on-chain usage. The
+// key difference is that one of the participants (the provider) commits to a sequence of random numbers
+// up-front using a hash chain. Users of the protocol then simply grab the next random number in the sequence.
+//
+// Setup: The provider P computes a sequence of N random numbers, x_i (i = 0...N-1):
+// x_{N-1} = random()
+// x_i = hash(x_{i + 1})
+// The provider commits to x_0 by posting it to the contract. Each random number in the sequence can then be
+// verified against the previous one in the sequence by hashing it, i.e., hash(x_i) == x_{i - 1}
+//
+// Request: To produce a random number, the following steps occur.
+// 1. The user draws a random number x_U, and submits h_U = hash(x_U) to this contract
+// 2. The contract remembers h_U and assigns it an incrementing sequence number i, representing which
+//    of the provider's random numbers the user will receive.
+// 3. The user submits an off-chain request (e.g. via HTTP) to the provider to reveal the i'th random number.
+// 4. The provider checks the on-chain sequence number and ensures it is > i. If it is not, the provider
+//    refuses to reveal the ith random number.
+// 5. The provider reveals x_i to the user.
+// 6. The user submits both the provider's revealed number x_i and their own x_U to the contract.
+// 7. The contract verifies hash(x_i) == x_{i-1} to prove that x_i is the i'th random number. The contract also checks that hash(x_U) == h_U.
+//    The contract stores x_i as the i'th random number to reuse for future verifications.
+// 8. If both of the above conditions are satisfied, the random number r = hash(x_i, x_U).
+//    (Optional) as an added security mechanism, this step can further incorporate the blockhash of the request transaction,
+//    r = hash(x_i, x_U, blockhash).
+//
+// This protocol has the same security properties as the 2-party randomness protocol above: as long as either
```

Comment:
I think it's not entirely true. I discussed it previously with @jayantk.

Reply:
yeah, i think you're right that it's somewhat easier to attack the hash function itself. I looked at a few papers on the topic and it seems like they do the same thing we're doing here though (no added random value per element). I figure let's go with this for the moment, and we can ask TOB at the audit stage what they think.